File: bit_manipulation/bit_tricks.cpp
#include <iostream>
using namespace std;

// Counts the number of one’s(set bits) in an integer(long/long long).
int getSetBits(uint32_t x) {
    return __builtin_popcount(x);
}

// Checks the Parity of a number. number has odd parity(odd number of set bits)?
int hasOddSetBits(uint32_t x) {
    return __builtin_parity(x);
}

// Counts the leading (left) number of zeros of the integer
int getLeadingZerosCount(uint32_t x) {
    return __builtin_clz(x);
    // return __builtin_clzll(x); // for long long
}

// Counts the trailing(right) number of zeros of the integer
int getTrailingZerosCount(uint32_t x) {
    return __builtin_ctz(x);
    // return __builtin_ctzll(x); // for long long
}

// highest power of 2 less than or equal to x
int highestPowerOf2(uint32_t x) {
    return 31 - __builtin_clz(x);
}

// isolate leftmost set bit / highest set bit / Most Significant Bit
int msb(int n) {
    return 1 << highestPowerOf2(n);
}

// isolate rightmost set bit / lowest set bit / Least Significant Bit
int lsb(int n) {
    return n == 0 ? 0 : n & -n;
}

int main() {
    // cout << highestPowerOf2(19) << endl;
    return 0;
}

/*************************************************************************/
// <----- leading zeroes ---->|     <--- trailing zeroes --->
// 00000000000000000000000000010011
// u-unsigned

// highest power of 2 less than or equal to x
int highestPowerOf2(int x) {
    // return 32 - getLeadingZerosCount(x) - 1;  // 0-based
    return 31 - getLeadingZerosCount(x);  // 0-based

    // return 32 - getLeadingZerosCount(x); // 1-based
}

File: bit_manipulation/bit_tricks.py

"""
2's complement
-x = ~x + 1
1. flip all bits
2. add 1 to it
"""


def lsb(n: int):
    """isolate rightmost set bit / lowest set bit / Least Significant Bit"""
    return 0 if n == 0 else n & -n


def highest_power_of_2(x: int):
    """highest power of 2 less than or equal to x"""
    return x.bit_length() - 1


def msb(n: int):
    """isolate leftmost set bit / highest set bit / Most Significant Bit"""
    return 1 << highest_power_of_2(n)


# highest power of 2 less than or equal to x
"""
from math import *
x = 16
print(bin(19))  # 10011

# Rank 1
k = x.bit_length() - 1
print(k)

# Rank 2 (Python only): using math
print(floor(log2(x)))

# Rank 3 (Python only)
LOG = 0
while (1 << (LOG + 1)) <= x:
    LOG += 1
print(LOG, 1 << LOG, bin(LOG))
"""


# isolate leftmost set bit / highest set bit / Most Significant Bit
"""
# faster
def msb(n: int):
    return 1 << (n.bit_length() - 1)
# slower
def msb(n: int):
    while n & (n-1):
        n = n & n - 1
    return n
"""
